src/briefcase/commands/build.py

```diff
@@ -14,16 +17,66 @@ def add_options(self, parser):
         self._add_update_options(parser, context_label=" before building")
         self._add_test_options(parser, context_label="Build")
 
-    def build_app(self, app: BaseConfig, **options):
+    def build_app(self, app: AppConfig, **options):
         """Build an application.
 
         :param app: The application to build
         """
         # Default implementation; nothing to build.
 
+    def verify_template(self, app: AppConfig):
+        """Verify the template satisfies the Command's requirements."""
+        self.logger.info("Checking template compatability...", prefix=app.app_name)
+
+        with self.input.wait_bar("Verifying Briefcase support..."):
+            # check against the "release" for each Version since versions
+            # such as "0.3.15.dev123" are considered older than "0.3.15"
+
+            if template_target := self.briefcase_target_version(app=app):
+                template_target = Version(template_target).release
+
+            if platform_target := self.oldest_compatible_briefcase:
+                platform_target = Version(platform_target).release
+
+            if platform_target and not template_target:
+                raise BriefcaseCommandError(
+                    f"""\
+The app template must declare a target version of Briefcase to confirm it is
+compatible with this version of Briefcase.
+
+The template's target Briefcase version must be {'.'.join(map(str, platform_target))} or later.
+
+Once the template is updated, run the create command:
+
+    $ briefcase create {self.platform} {self.output_format}
+"""
+                )
+
+            current_version = Version(briefcase.__version__).release
+            if (platform_target and platform_target > template_target) or (
+                template_target and template_target > current_version
+            ):
```

Comment:
This should be an `elif platform_target != template_target`. There's no need for the actual Briefcase version to be involved, and it's not a > comparison - we're looking for a "Briefcase version as epoch" marker. For this change to Java, the template_target will be `0.3.15`, and it will stay that way until we make another backwards incompatible change to the Android template format.

Reply:
I already discussed this [here](https://github.com/beeware/briefcase/pull/1271#issuecomment-1563470051) and [here](https://github.com/beeware/briefcase/pull/1271#issuecomment-1557963948).

Reply:
this has definitely been a sticking point...i guess i won't say too much more except these approaches are effectively the same...with slightly different implementations.

I like the idea of an epoch....but I think using the Briefcase version is confusing....because people will think you're literally talking about the current version of Briefcase.

I also like the idea of supporting a version inside the template that's independent of the platform version support declaration in Briefcase. This allows third party templates to specify an arbitrary level of support with Briefcase. If we limit this to Briefcase's internal platform version, then we're limiting the number of potential breaking changes third party can protect themselves from....which could be missed breaking changes or even just nuances among versions.

At this point, I'm starting to feel we may just need an executive decision....if only because I'm a bit exhausted going round and round. I'm prepared to implement the epoch versioning or even something else. (I'm still working on promoting the check to all Commands since that's pretty impactful to the tests.)

Reply:
In which case, I'm going to put on my BDFN hat and say "epoch-style numbering, using the Briefcase version as the epoch marker". 

Yes, there's a source of confusion in that a 0.3.16 template will have marker saying it is for 0.3.15. However, as this discussion has revealed, *any* discussion of compatibility and version numbers is going to be confusing; picking an identifier that has a clear interpretation (vs a new "epoch" number that will need it's own explanation) is about as good as we're going to get, IMHO. 

It's also not especially user-visible - it's a value in `briefcase.toml`, which isn't something that run-of-the-mill users will ever need to see. This is a "confusing detail" that will mostly affect Briefcase maintainers, not the outside world.

Using the 0.3.15 version is also implicit documentation of which release notes you need to look at for a migration guide.

I acknowledge that there *could* be a need for a third-party template provider to lock against a different Briefcase version. However, we don't have any particular evidence that this requirement exists - it's mostly speculation and trying to foretell the future. On that basis, I'm comfortable living with the limitation until such time as we can clearly identify a specific use-case. *If* such a requirement were to develop, I can imagine adding an additional template option (or command line option) to ignore template version checks, or impose a *different* version check; however, I think we should cross that bridge when (and if) we come to it.